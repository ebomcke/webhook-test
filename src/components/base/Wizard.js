import React, { Component } from 'react';
import { StepGroup, Form } from 'semantic-ui-react';

class Wizard extends Component {
  constructor(props) {
    super(props);
    this.state = {
      activeStep: 0,
    };
  }

  nextStep = async () => {
    const { steps, onComplete } = this.props;
    const activeStep = this.state.activeStep;
    if (activeStep < steps.length - 1) {
      this.setState({
        activeStep: activeStep + 1,
      });
    } else {
      await onComplete();
    }
  }

  render() {
    const { steps } = this.props;
    const stepItems = steps.map((step, index) => ({
      key: step.key,
      title: step.title,
      icon: step.icon,
      description: step.description,
      active: index === this.state.activeStep,
      disabled: index > this.state.activeStep,
    }));
    const activeStep = steps[this.state.activeStep];
    const nextButton = () => (
      <Form.Button
        wizardIdentifier='next-done-button'
        onClick={this.nextStep}
        disabled={activeStep.nextDisabled && activeStep.nextDisabled()}
      >
        {this.state.activeStep === steps.length - 1 ? 'Done' : 'Next'}
      </Form.Button>
    );
    return (
      <div>
        <StepGroup fluid items={stepItems} />
        {activeStep.render(nextButton)}
      </div>
    );
  }
}

export default Wizard;
