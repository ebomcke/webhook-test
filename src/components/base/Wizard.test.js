import React from 'react';
import Wizard from './Wizard';
import renderer from 'react-test-renderer';

const generateStep = key => ({
  key: key,
  title: `${key} - title`,
  description: `${key} - description`,
  render: nextButton => (
    <div>
      <p>
        This is where the content for {key} goes, below is the next/done button.
      </p>
      {nextButton()}
    </div>
  ),
});

describe('Rendering tests', () => {
  test('Single step', () => {
    const steps = [generateStep('test')];
    const wizard = renderer.create(<Wizard steps={steps} />);
    expect(wizard.toJSON()).toMatchSnapshot();
  });

  test('Multiple steps', () => {
    const steps = [
      generateStep('item-1'),
      generateStep('item-2'),
      generateStep('item-3'),
    ];
    const wizard = renderer.create(<Wizard steps={steps} />);
    expect(wizard.toJSON()).toMatchSnapshot();
  });
});

describe('Interactions tests', () => {
  const steps = [
    generateStep('item-1'),
    generateStep('item-2'),
    generateStep('item-3'),
  ];
  let wizard;
  const onComplete = jest.fn();

  beforeEach(() => {
    wizard = renderer.create(<Wizard steps={steps} onComplete={onComplete} />);
  });

  afterEach(() => {
    onComplete.mockClear();
  });

  test('Clicking on the next button shows next step', () => {
    wizard.root
      .findByProps({ wizardIdentifier: 'next-done-button' })
      .props.onClick();
    expect(wizard.toJSON()).toMatchSnapshot();
  });

  test('Clicking on the done button calls onComplete', () => {
    wizard.root
      .findByProps({ wizardIdentifier: 'next-done-button' })
      .props.onClick();
    wizard.root
      .findByProps({ wizardIdentifier: 'next-done-button' })
      .props.onClick();
    expect(wizard.toJSON()).toMatchSnapshot();
    wizard.root
      .findByProps({ wizardIdentifier: 'next-done-button' })
      .props.onClick();
    expect(onComplete.mock.calls.length).toBe(1);
  });
});
