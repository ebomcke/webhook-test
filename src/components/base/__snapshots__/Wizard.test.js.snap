// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Interactions tests Clicking on the done button calls onComplete 1`] = `
<div>
  <div
    className="ui fluid steps"
  >
    <div
      className="step"
      onClick={[Function]}
    >
      <div
        className="content"
      >
        <div
          className="title"
        >
          item-1 - title
        </div>
        <div
          className="description"
        >
          item-1 - description
        </div>
      </div>
    </div>
    <div
      className="step"
      onClick={[Function]}
    >
      <div
        className="content"
      >
        <div
          className="title"
        >
          item-2 - title
        </div>
        <div
          className="description"
        >
          item-2 - description
        </div>
      </div>
    </div>
    <div
      className="active step"
      onClick={[Function]}
    >
      <div
        className="content"
      >
        <div
          className="title"
        >
          item-3 - title
        </div>
        <div
          className="description"
        >
          item-3 - description
        </div>
      </div>
    </div>
  </div>
  <div>
    <p>
      This is where the content for 
      item-3
       goes, below is the next/done button.
    </p>
    <div
      className="field"
    >
      <button
        className="ui button"
        onClick={[Function]}
        wizardIdentifier="next-done-button"
      >
        Done
      </button>
    </div>
  </div>
</div>
`;

exports[`Interactions tests Clicking on the next button shows next step 1`] = `
<div>
  <div
    className="ui fluid steps"
  >
    <div
      className="step"
      onClick={[Function]}
    >
      <div
        className="content"
      >
        <div
          className="title"
        >
          item-1 - title
        </div>
        <div
          className="description"
        >
          item-1 - description
        </div>
      </div>
    </div>
    <div
      className="active step"
      onClick={[Function]}
    >
      <div
        className="content"
      >
        <div
          className="title"
        >
          item-2 - title
        </div>
        <div
          className="description"
        >
          item-2 - description
        </div>
      </div>
    </div>
    <div
      className="disabled step"
      onClick={[Function]}
    >
      <div
        className="content"
      >
        <div
          className="title"
        >
          item-3 - title
        </div>
        <div
          className="description"
        >
          item-3 - description
        </div>
      </div>
    </div>
  </div>
  <div>
    <p>
      This is where the content for 
      item-2
       goes, below is the next/done button.
    </p>
    <div
      className="field"
    >
      <button
        className="ui button"
        onClick={[Function]}
        wizardIdentifier="next-done-button"
      >
        Next
      </button>
    </div>
  </div>
</div>
`;

exports[`Rendering tests Multiple steps 1`] = `
<div>
  <div
    className="ui fluid steps"
  >
    <div
      className="active step"
      onClick={[Function]}
    >
      <div
        className="content"
      >
        <div
          className="title"
        >
          item-1 - title
        </div>
        <div
          className="description"
        >
          item-1 - description
        </div>
      </div>
    </div>
    <div
      className="disabled step"
      onClick={[Function]}
    >
      <div
        className="content"
      >
        <div
          className="title"
        >
          item-2 - title
        </div>
        <div
          className="description"
        >
          item-2 - description
        </div>
      </div>
    </div>
    <div
      className="disabled step"
      onClick={[Function]}
    >
      <div
        className="content"
      >
        <div
          className="title"
        >
          item-3 - title
        </div>
        <div
          className="description"
        >
          item-3 - description
        </div>
      </div>
    </div>
  </div>
  <div>
    <p>
      This is where the content for 
      item-1
       goes, below is the next/done button.
    </p>
    <div
      className="field"
    >
      <button
        className="ui button"
        onClick={[Function]}
        wizardIdentifier="next-done-button"
      >
        Next
      </button>
    </div>
  </div>
</div>
`;

exports[`Rendering tests Single step 1`] = `
<div>
  <div
    className="ui fluid steps"
  >
    <div
      className="active step"
      onClick={[Function]}
    >
      <div
        className="content"
      >
        <div
          className="title"
        >
          test - title
        </div>
        <div
          className="description"
        >
          test - description
        </div>
      </div>
    </div>
  </div>
  <div>
    <p>
      This is where the content for 
      test
       goes, below is the next/done button.
    </p>
    <div
      className="field"
    >
      <button
        className="ui button"
        onClick={[Function]}
        wizardIdentifier="next-done-button"
      >
        Done
      </button>
    </div>
  </div>
</div>
`;
